lib/pbench/server/database/models/datasets.py

```diff
@@ -779,8 +779,25 @@ def is_key_path(key: str, valid: List[str], metalog_key_ok: bool = False) -> boo
     def getvalue(
         dataset: Dataset, key: str, user_id: Optional[str] = None
     ) -> Optional[JSON]:
-        """Returns the value of the specified key, which may be a dotted
-        hierarchical path (e.g., "server.deleted").
+        """Returns the value of the specified metadata key.
+
+        A metadata key may be a dotted hierarchical path, like "server.deleted"
+        or "dataset.metalog.pbench.script".
+
+        This code assumes that any necessary syntax checks have been made by a
+        caller. Normally, user key update is restricted to a specific set of
+        keys and namespaces (first level key like "server" or "global"), and
+        keys are lowercase and alphanumeric.
+
+        In the "dataset.metalog" sub-namespace, keys originate from a Pbench
+        Agent "metadata.log" config file, which may include non-alphanumeric
+        keys such as "iterations/1-tcp_stream-64B-1i". Pbench Server APIs may
```

Comment:
If someone were to (somehow) set a metadata key (lets say, `"global.mykey"`) to a JSON blob like
```json
{
  "level.one": {
    "level.two": "my_value"
}
```
and then lobbed a query or selection for `?metadata="global.mykey.level.one.level.two"` then it would all "just work", yes?  Because we would have stored it as
```javascript
{
  "global": {
    "mykey": {
      "level": {
        "one": {
          "level": {
            "two": "my_value"
          }
        }
      }
    }
  }
}
```
and the dot-parsing would find it.

Reply:
This is one of those "really good questions," which actually has little to do with the intent of this PR but possibly could/should/might be solved along with it. There are now two ways to set metadata keys.

1. `PUT /upload/{name}?metadata=global.mykey.level.one.level.two:my_value`, the interpretation of which is completely up to us. We'll separate the key path on the "." characters and construct the final JSON you showed.
2. `PUT /datasets/{id}/metadata {"metadata": ...}` which gives the client complete control over the structure of the JSON. And if the body is `{"metadata": {"global": {"mykey": {"level.one": {"level.two": "my_value"}}}}}` that's *totally* different. Note that `{"metadata": {"global.mykey": { ... }}}` will be rejected because "global.mykey" isn't a valid root namespace: but we don't care about the lower keys. You'll get exactly what you specified, and a `GET` on `global.mykey.level.one.level.two` will return `None` because there is no such key. I'm honestly not sure whether this is a bug. They won't be able to GET `global.mykey.level` because it'll look for `{"global": {"mykey": {"level": ... }}}` and that's not what's there... but a `GET` on `global.mykey` will return the JSON block they supplied.
 
I'm currently doing *no* syntax checking within a JSON blob. We know it's valid JSON because the validation would have caught a structural problem, and I never really worried about it beyond that: but `PUT /datasets/{id}/metadata {"global": {"Weird@key#1": true}}` will work. And of course you can always get it back with `GET /datasets/{id}/metadata?metadata=global`, but `GET /datasets/{id}/metadata?metadata=global.Weird@key#1` won't work. One alternative would be to scan through a JSON object to validate each key and reject any key that doesn't conform to the basic lowercase alphanumeric pattern. I don't think I'd want to try to mangle the payload to break it up: but we could reject any structure with non-alphanumeric keys.

Of course the other alternative is to leave it as it is. Yeah, they can specify a JSON structure they can't pull back out except as a block; weird, but maybe not the worst thing in the world. That is, we validate the primary DB key (the namespace value), and you can always `GET /datasets/{id}/metadata?metadata=global` to get back everything. Hmm.